File: graph/unionfind.cpp
class UnionFind{
	vector<int> par;
	vector<int> rank;
public:
	UnionFind(int max_n):par(max_n),rank(max_n){}
	void init(int n){
		for(int i=0;i<n;i++){
			par[i]=i;
			rank[i]=0;
		}
	}
	int find(int x){
		if(par[x]==x)return x;
		return par[x]=find(par[x]);
	}
	void unite(int x,int y){
		x=find(x);
		y=find(y);
		if(rank[x]<rank[y]){
			par[x]=y;
		}else{
			par[y]=x;
			if(rank[x]==rank[y])rank[x]++;
		}
	}
	bool same(int x,int y){
		return find(x)==find(y);
	}
};

File: rust/src/datastructure/union_find.rs
pub struct UnionFind {
    par: Vec<usize>,
    size: Vec<usize>,
}

impl UnionFind {
    pub fn new(n: usize) -> Self {
        let mut uf = UnionFind {
            par: Vec::with_capacity(n),
            size: vec![1; n],
        };
        for i in 0..n {
            uf.par.push(i);
        }
        return uf;
    }
    pub fn find(&mut self, u: usize) -> usize {
        if self.par[u] != u {
            let p = self.par[u];
            self.par[u] = self.find(p);
            return self.par[u];
        }
        return u;
    }
    pub fn size(&mut self, u: usize) -> usize {
        let u = self.find(u);
        return self.size[u];
    }
    pub fn same(&mut self, u: usize, v: usize) -> bool {
        return self.find(u) == self.find(v);
    }
    pub fn unite(&mut self, u: usize, v: usize) {
        let u = self.find(u);
        let v = self.find(v);
        if u == v {
            return;
        }
        if self.size[u] < self.size[v] {
            self.par[u] = v;
            self.size[v] += self.size[u];
        } else {
            self.par[v] = u;
            self.size[u] += self.size[v];
        }
    }
}
